neps/nep-0399.md

```diff
@@ -0,0 +1,461 @@
+in NEPs like https://github.com/near/NEPs/pull/460.
+
+Risk of DB corruption slightly increases, and it becomes harder to replay blocks on chain. While `Trie` entries are
+essentially immutable (in fact, value for each key is
+unique, because key is a value hash), `FlatStorage` is read-modify-write, because values for the same `TrieKey` can be
+completely different. We believe that such flat mapping is reasonable to maintain anyway, as for newly discovered state
+sync idea. But if some change was applied incorrectly, we may have to recompute the whole flat storage, and for block
+hashes before flat head we can't access flat storage at all.
+
+Though Flat Storage significantly reduces amount of storage reads, we have to keep it up-to-date, which results in 1
+extra disk write for changed key, and 1 auxiliary disk write + removal for each FSD. Disk requirements also slightly
+increase. We think it is acceptable, because actual disk writes are executed in background and are not a bottleneck
+for block processing. For storage write in general Flat Storage is even a net improvement, because it removes
+necessity to traverse changed nodes during write execution ("reads-for-writes"), and we can apply optimizations
+there (see "Storage Writes" section).
+
+Implementing FlatStorage will require a lot of engineering effort and introduce code that will make the codebase more 
+complicated. In particular, we had to extend `RuntimeAdapter` API with flat storage-related method after thorough
+considerations. We are confident that FlatStorage will bring a lot of performance benefit, but we can only measure the exact
+improvement after the implementation. We may find that the benefit FlatStorage brings is not 
+worth the effort, but it is very unlikely.
+
+It will make the state rollback harder in the future when we enable challenges in phase 2 of sharding.
+When a challenge is accepted and the state needs to be rolled back to a previous block, the entire flat state needs to 
+be rebuilt, which could take a long time. Alternatively, we could postpone garbage collection of deltas and add support 
+of applying them backwards.
+
+Speaking of new sharding phases, once nodes are no longer tracking all shards, Flat Storage must have support for adding
+or removing state for some specific shard. Adding new shard is a tricky but natural extension of catchup process. Our 
+current approach for removal is to iterate over all entries in `DBCol::FlatState` and find out for each trie key to
+which shard it belongs to. We would be happy to assume that each shard is represented by set of
+contiguous ranges in `DBCol::FlatState` and make removals simpler, but this is still under discussion.
+
+Last but not least, resharding is not supported by current implementation yet.
+
+## Unresolved Issues
+
+As we discussed in previous sections, there are still unanswered questions around how the new cost model for storage
```

Comment:
Maybe this section needs to be updated, since the issues listed are discussed in the NEP already.

Reply:
In fact it's completely covered by Drawbacks. Removed it as optional, added "Future possibilities" instead.